src/loan.rs

```diff
@@ -1,28 +1,80 @@
-use elements::hashes::{sha256d, Hash};
-use elements::opcodes::all::*;
-use elements::script::Builder;
+use elements::confidential::{Asset, AssetBlindingFactor, ValueBlindingFactor};
+use elements::encode::serialize;
 use elements::secp256k1_zkp::rand::{CryptoRng, RngCore};
-use elements::secp256k1_zkp::{Secp256k1, SecretKey, Signature, Signing, Verification, SECP256K1};
+use elements::secp256k1_zkp::{Secp256k1, SecretKey, Signing, Verification, SECP256K1};
 use elements::sighash::SigHashCache;
 use elements::{
-    Address, AddressParams, AssetId, OutPoint, Script, SigHashType, Transaction, TxIn, TxInWitness,
-    TxOut, TxOutSecrets,
+    Address, AddressParams, AssetId, OutPoint, SigHashType, Transaction, TxIn, TxInWitness, TxOut,
+    TxOutSecrets,
 };
-use secp256k1_zkp::rand::thread_rng;
+use elements_miniscript::descriptor::{CovSatisfier, ElementsTrait};
+use elements_miniscript::miniscript::satisfy::After;
+use elements_miniscript::{Descriptor, DescriptorTrait};
 use secp256k1_zkp::{SurjectionProof, Tag};
+use std::collections::HashMap;
 use std::future::Future;
+use std::str::FromStr;
 
 #[cfg(test)]
 mod protocol_tests;
 mod stack_simulator;
 
+static COVENANT_SK: Lazy<SecretKey> = Lazy::new(|| {
+    SecretKey::from_str("cc5417e929f7756df9a599715ad0780cea75659279cd4e2c0a19adb6339d7011")
+        .expect("is a valid key")
+});
+static COVENANT_PK: Lazy<PublicKey> = Lazy::new(|| {
+    PublicKey::from_private_key(
+        &SECP256K1,
+        &PrivateKey {
+            compressed: true,
+            network: Network::Regtest,
+            key: *COVENANT_SK,
+        },
+    )
+});
+
+/// Generate the miniscript descriptor of the collateral output.
+///
+/// It defines a "liquidation branch" which allows the lender to claim
+/// all the collateral for themself if the `timelock` expires. The
+/// lender must identify themself by providing a signature on
+/// `lender_pk`.
+///
+/// It also defines a "repayment branch" which allows the borrower to
+/// repay the loan to reclaim the collateral. The borrower must
+/// identify themself by providing a signature on `borrower_pk`. To
+/// ensure that the borrower does indeed repay the loan, the script
+/// will check that the spending transaction has the
+/// `repayment_output` as vout 0.
+///
+/// The first element of the covenant descriptor is the shared
+/// `covenant_pk`, which is only used to verify that the transaction
+/// data on the witness stack matches the transaction which triggered
+/// the call.
+fn loan_descriptor(
+    borrower_pk: PublicKey,
+    lender_pk: PublicKey,
+    timelock: u64,
+    repayment_output: TxOut,
+) -> Result<Descriptor<PublicKey>> {
+    let repayment_output = serialize(&repayment_output).to_hex();
+    let desc = Descriptor::<elements::bitcoin::PublicKey>::from_str(
+        &(format!(
+            "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
+            *COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
```

Comment:
This is the only place where you use `COVENANT_PK` and it is being `Diplay`ed. You could store the PK as a hex string right away and verify with a test that it is indeed the PK to the SK.

Would reduce indirection a bit and do less CPU cycles :)

Reply:
Also, it would get rid of the `network: Network::Regtest,` above which scared the shit out of me until I remembered that public keys don't have a network-specific encoding and this is only there for WIF I think.